nutils/evaluable.py

```diff
@@ -4217,6 +4199,91 @@ def replace_arguments(value, arguments):
     assert equalshape(value.shape, v.shape)
     return v
 
+def einsum(fmt, *args, **dims):
+  '''Multiply and/or contract arrays via format string.
+
+  The format string consists of a comma separated list of axis labels, followed
+  by ``->`` and the axis labels of the return value. For example, the following
+  swaps the axes of a matrix:
+
+  >>> einsum('ij->ji', ones([2,3]))
+  nutils.evaluable.Transpose<f:3,2>
+
+  Axis labels that no not occur in the return value are summed. For example,
+  the following performs a dot product of three matrices:
+
+  >>> einsum('ij,jk,kl->il', ones([2,3]), ones([3,4]), ones([4,5]))
+  nutils.evaluable.Sum<f:2,5>
+
+  In case the dimension of the input and output arrays may vary, a variable
+  length axes group can be denoted by a capital. Its length is automatically
+  established based on the dimension of the input arrays. The following example
+  performs a tensor product of an array and a vector:
+
+  >>> einsum('A,i->Ai', ones([2,3,4]), ones([5]))
+  nutils.evaluable.Multiply<f:2,3,4,5>
+
+  The format string may contain multiple variable length axes groups, but their
+  lengthts must be resolvable from left to right. In case this is not possible,
+  lengths may be specified as keywords arguments.
+
+  >>> einsum('AjB,i->AijB', ones([2,3,4]), ones([5]), B=1)
+  nutils.evaluable.Multiply<f:2,5,3,4>
+  '''
+
+  sin, sout = fmt.split('->')
+  sin = sin.split(',')
+
+  if len(sin) != len(args):
+    raise ValueError('number of arguments does not match format string')
+
+  if any(len(s) != len(set(s)) for s in (*sin, sout)):
+    raise ValueError('internal repetitions are not supported')
+
+  for c in 'abcdefghijklmnopqrstuvwxyz':
+    dims.setdefault(c, 1) # lowercase characters default to single dimension
+
+  for s, arg in zip(sin, args):
+    unknown = [c for c in s if c not in dims]
+    if unknown:
+      if len(unknown) > 1:
+        raise ValueError('cannot establish length of variable groups {}'.format(', '.join(unknown)))
+      E, = unknown
+      dims[E] = arg.ndim - builtins.sum(dims[c] for c in s if c != E)
+    else:
+      if arg.ndim != builtins.sum(dims[c] for c in s):
+        raise ValueError('argument dimensions are inconsistent with format string')
+
+  neg = [c for c, n in dims.items() if n < 0]
+  if neg:
+    raise ValueError('group {} evaluates to negative dimension'.format(', '.join(neg)))
```

Comment:
I think that the error message 'group evaluates to negative dimension' is a bit cryptic. Perhaps 'argument dimensions are inconsistent', as is raised for arguments with invalid dimensions without unknowns, is easier to understand. Suggestion:

```python
  for s, arg in zip(sin, args):
    unknown = [c for c in s if c not in dims]
    known_dim = builtins.sum(dims.get(c, 0) for c in s)
    if unknown:
      if len(unknown) > 1:
        raise ValueError('cannot establish length of variable groups {}'.format(', '.join(unknown)))
      E, = unknown
      dims[E] = dimE = max(0, arg.ndim - known_dim)
      known_dim += dimE
    if arg.ndim != known_dim:
      raise ValueError('argument dimensions are inconsistent with format string')
```
If a user passes a negative dimension to argument `dims`, this could be a different error, for example 'all predefined dimensions must be non-negative'.

Reply:
fixed (slightly different implementation)